File: HW 6 (Multithreading)/multithreading/src/main/java/com/example/multithreading/ContactListActivity.java
package com.example.multithreading;

import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.SearchView;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.multithreading.Threads.DataRepository;
import com.example.multithreading.Threads.FutureRepository;
import com.example.multithreading.Threads.HandlerRepository;
import com.example.multithreading.Threads.MultithreadingRegimeActivity;
import com.example.multithreading.Threads.MultithreadingRegimeManager;
import com.example.multithreading.Threads.RXJavaRepository;
import com.example.multithreading.Threads.RepositoryListener;
import com.example.multithreading.database.AppDatabase;
import com.example.multithreading.database.Contact;
import com.example.multithreading.database.ContactDao;
import com.google.android.material.floatingactionbutton.FloatingActionButton;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Executor;
import java.util.concurrent.Executors;
import java.util.concurrent.ThreadPoolExecutor;

public class ContactListActivity extends AppCompatActivity {

    public static final String CONTACT_FOR_EDITING = "CONTACT_FOR_EDITING";
    private RecyclerView contactsListRecycleView;
    private ContactsAdapter adapter;
    private TextView noContacts;
    private List<Contact> contactsList = new ArrayList<>();

    private ContactDao contactDao;
    private DataRepository dataRepository;
    private MultithreadingRegimeManager regimeManager;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_contacts_list);

        contactsListRecycleView = findViewById(R.id.recycleView);
        noContacts = findViewById(R.id.noContacts);
        FloatingActionButton floatingButton = findViewById(R.id.floatingButton_addContact);
        Toolbar toolbar = findViewById(R.id.toolbarMain);
        setSupportActionBar(toolbar);
        SearchView search = findViewById(R.id.search);
        floatingButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                startAddContactActivity();
            }
        });

        AppDatabase appDatabase = AppDatabase.getAppDatabaseInstance(this);
        contactDao = appDatabase.getContactDao();
        regimeManager = MultithreadingRegimeManager.getRegimeManager(ContactListActivity.this);
        getMultithreadingRegime();

        contactsListRecycleView.setAdapter(new ContactsAdapter(contactsList, new ContactsAdapter.OnContactClickListener() {
            @Override
            public void onContactClick(Contact contact) {
                startEditContactActivity(contact);
            }
        }));
        contactsListRecycleView.setLayoutManager(new LinearLayoutManager(this, RecyclerView.VERTICAL, false));

        adapter = (ContactsAdapter) contactsListRecycleView.getAdapter();
    }

    @Override
    protected void onStart() {
        super.onStart();
        loadContactList();
    }

    private void startAddContactActivity() {
        Intent intent = new Intent(ContactListActivity.this, AddContactActivity.class);
        startActivity(intent);
    }

    private void startEditContactActivity(Contact contact) {
        Intent intent = new Intent(ContactListActivity.this, EditContactActivity.class);
        intent.putExtra(CONTACT_FOR_EDITING, contact.getId());
        startActivity(intent);
    }

    private void startMultithreadingRegimeActivity() {
        Intent intent = new Intent(ContactListActivity.this, MultithreadingRegimeActivity.class);
        startActivity(intent);
    }

    private void checkEmpty(List<Contact> list) {
        if (list.isEmpty()) {
            contactsListRecycleView.setVisibility(View.GONE);
            noContacts.setVisibility(View.VISIBLE);
        } else {
            contactsListRecycleView.setVisibility(View.VISIBLE);
            noContacts.setVisibility(View.GONE);
        }
    }

    private void loadContactList() {
        dataRepository.getAll(new RepositoryListener<List<Contact>>() {
            @Override
            public void onDataReceived(List<Contact> contactList) {
                if (adapter != null) {
                    adapter.items = contactList;
                    adapter.notifyDataSetChanged();
                    checkEmpty(adapter.items);
                }
            }
        });

    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.contact_list_activity_menu, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        int itemId = item.getItemId();
        switch (itemId) {  // я сделела здесь switch, а не if для того, чтобы легче было добавлять другие функции в меню, если они появятся.
            case R.id.multithreadingRegime:
                startMultithreadingRegimeActivity();
                break;
            default:
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    private void getMultithreadingRegime() {
        int regimeId = regimeManager.loadMultithreadingRegime();
        switch (regimeId) {
            case R.id.executorHandler:
                dataRepository = new HandlerRepository(contactDao, (ThreadPoolExecutor) Executors.newFixedThreadPool(2));
                break;
            case R.id.executorFuture:
                dataRepository = new FutureRepository(contactDao, (Executor) getMainExecutor(),
                        (ThreadPoolExecutor) Executors.newFixedThreadPool(2));
                break;
            case R.id.rxJava:
                dataRepository = new RXJavaRepository(contactDao);
                break;
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        dataRepository.close();
    }

    // ADAPTER
    private static class ContactsAdapter extends RecyclerView.Adapter<ContactsAdapter.ContactsViewHolder> {
        public interface OnContactClickListener {
            void onContactClick(Contact contact);
        }

        private List<Contact> items;
        private final OnContactClickListener contactListener;

        public ContactsAdapter(List<Contact> items, OnContactClickListener contactListener) {
            this.items = items;
            this.contactListener = contactListener;
        }

        @NonNull
        @Override
        public ContactsViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            View view = LayoutInflater
                    .from(parent.getContext())
                    .inflate(R.layout.item_contact, parent, false);
            return new ContactsViewHolder(view);
        }

        @Override
        public void onBindViewHolder(@NonNull ContactsViewHolder holder, int position) {
            holder.bind(items.get(position), contactListener);
        }

        @Override
        public int getItemCount() {
            return items != null ? items.size() : 0;
        }

        // VIEWHOLDER
        private static class ContactsViewHolder extends RecyclerView.ViewHolder {
            private ImageView itemIcon;
            private TextView itemPersonName;
            private TextView itemPersonContact;

            public ContactsViewHolder(@NonNull View itemView) {
                super(itemView);

                itemIcon = itemView.findViewById(R.id.item_icon);
                itemPersonName = itemView.findViewById(R.id.item_personName);
                itemPersonContact = itemView.findViewById(R.id.item_personContact);
            }

            public void bind(final Contact contact, final OnContactClickListener contactListener) {

                itemPersonName.setText(contact.getPersonName());
                itemPersonContact.setText(contact.getContactDetails());

                switch (contact.getContactType()) {
                    case PHONE:
                        itemIcon.setImageResource(R.drawable.contact_phone);
                        break;
                    case EMAIL:
                        itemIcon.setImageResource(R.drawable.contact_email);
                        break;
                    default:
                        break;
                }

                itemView.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View view) {
                        contactListener.onContactClick(contact);
                    }
                });
            }
        }
    }
}

File: HW 9 (ContentProvider, ContentResolver)/contentprovider/src/main/java/com/example/contentprovider/ContactListActivity.kt
package com.example.contentprovider

import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.contentprovider.database.AppDatabase
import com.example.contentprovider.database.CONTACT_TYPE.EMAIL
import com.example.contentprovider.database.CONTACT_TYPE.PHONE
import com.example.contentprovider.database.Contact
import com.example.contentprovider.database.ContactDao
import kotlinx.android.synthetic.main.activity_contact_list.*
import kotlinx.android.synthetic.main.item_contact.view.*

internal const val CONTACT_FOR_EDITING: String = "CONTACT_FOR_EDITING"

class ContactListActivity : AppCompatActivity() {

    private lateinit var contactDao: ContactDao
    private lateinit var contactsList: List<Contact>

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_contact_list)

        setSupportActionBar(toolbarMain)

        val appDatabase = AppDatabase.getAppDatabaseInstance(this)
        contactDao = appDatabase.getContactDao()
        contactsList = contactDao.getAll()

        recycleView.adapter = ContactsAdapter(contactsList, object : OnContactClickListener {
            override fun onContactClick(contact: Contact) {
                startEditContactActivity(contact)
            }
        })
        recycleView.layoutManager = LinearLayoutManager(this, RecyclerView.VERTICAL, false)

        floatingButton.setOnClickListener { startAddContactActivity() }
    }

    private fun startAddContactActivity() {
        val intent = Intent(this, AddContactActivity::class.java)
        startActivity(intent)
    }

    private fun startEditContactActivity(contact: Contact) {
        val intent = Intent(this, EditContactActivity::class.java)
        intent.putExtra(CONTACT_FOR_EDITING, contact.id)
        startActivity(intent)
    }

    private fun checkEmpty(list: List<Contact>) {
        if (list.isEmpty()) {
            recycleView.visibility = View.GONE
            noContacts.visibility = View.VISIBLE
        } else {
            recycleView.visibility = View.VISIBLE
            noContacts.visibility = View.GONE
        }
    }

    private fun loadContactList() {
        val adapter = recycleView.adapter as? ContactsAdapter
        adapter?.items = contactDao.getAll()
        adapter?.notifyDataSetChanged()
        adapter?.items?.let { checkEmpty(it) }
    }

    override fun onStart() {
        super.onStart()
        loadContactList()
    }

    // ADAPTER
    class ContactsAdapter(
            internal var items: List<Contact>,
            private val contactListener: OnContactClickListener
    )
        : RecyclerView.Adapter<ContactsAdapter.ContactsViewHolder>() {

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ContactsViewHolder {
            val view = LayoutInflater
                    .from(parent.context)
                    .inflate(R.layout.item_contact, parent, false)
            return ContactsViewHolder(view)
        }

        override fun onBindViewHolder(holder: ContactsViewHolder, position: Int) {
            holder.bind(items[position], contactListener)
        }

        override fun getItemCount(): Int {
            return items.size
        }

        // VIEWHOLDER
        class ContactsViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

            internal fun bind(contact: Contact, contactListener: OnContactClickListener) {
                itemView.itemPersonName.text = contact.personName
                itemView.itemPersonContact.text = contact.contactDetails
                when (contact.contactType) {
                    PHONE -> itemView.itemIcon.setImageResource(R.drawable.contact_phone)
                    EMAIL -> itemView.itemIcon.setImageResource(R.drawable.contact_email)
                }
                itemView.setOnClickListener { contactListener.onContactClick(contact) }
            }
        }
    }
}